DOCSP-15915 crud bulk operations (#81)
mongodb/docs-java: source/fundamentals/crud/bulk.txt

Delete Operation
~~~~~~~~~~~~~~~~

To perform a delete operation, create a ``DeleteOneModel`` or a
``DeleteManyModel`` specifying a query filter for documents you want to
delete. 

The ``DeleteOneModel`` deletes the first document that matches your query
filter and the ``DeleteManyModel`` deletes all the documents that
match your query filter. 

.. note::

   When performing a ``bulkWrite()``, the ``DeleteOneModel`` and
   ``DeleteOneModel`` do not delete any documents if there are no matches
   to your query filter. 

Example
```````

The following example creates a ``DeleteOneModel`` to delete
a document where the ``_id`` is "1":

.. literalinclude:: /includes/fundamentals/code-snippets/BulkWrite.java
   :language: java
   :dedent:
   :start-after: begin deleteDocumentsExample
   :end-before: end deleteDocumentsExample

For additional information on the classes mentioned in this
section, see the following resources: 

- :java-core-api:`DeleteOneModel <com/mongodb/client/model/DeleteOneModel.html>` API Documentation
- :java-core-api:`DeleteManyModel <com/mongodb/client/model/DeleteManyModel.html>` API Documentation

.. _orderOfExecution:

Order of Execution
------------------

> The ``bulkWrite()`` method also optionally takes ``BulkWriteOptions`` as
> a second parameter to specify if you want to execute the bulk operations

Reviewer: suggestion: This wording feels a little off, could just be me.

```suggestion
The ``bulkWrite()`` method accepts an optional``BulkWriteOptions`` as
a second parameter to specify if you want to execute the bulk operations
```

update multiple documents, and delete one document in your collection,
you use the ``insertOne()``, ``updateMany()`` and ``deleteOne()`` methods. 

Instead of using three different methods to perform three tasks, you can
use bulk operations, which uses one method to perform three tasks. Bulk
operations consist of a large number of CRUD operations and can perform
them using the ``bulkWrite()`` method. 

Performing Bulk Operations
--------------------------

The ``bulkWrite()`` method accepts a list of bulk operations in the
format of ``WriteModel`` documents. A ``WriteModel`` represents one of
the CRUD operations. The method also optionally takes
``BulkWriteOptions`` as a second parameter to specify the order you want
the operations to be perform.

The following snippet creates a ``List`` of bulk operations to pass into
the ``bulkWrite()`` method: 

.. literalinclude:: /includes/fundamentals/code-snippets/BulkWrite.java
   :language: java
   :dedent:
   :start-after: begin bulkOpsList
   :end-before: end bulkOpsList

The following sections show how to create and use each type of
``WriteModel`` document. The examples in each section contain the

> following documents in a collection:

Reviewer: suggestion: Add a bit more introductory material for WriteModel. As is, it feels a bit abrupt.

.. code-block:: json

   { "_id": 1 }
   { "_id": 2 }

For additional information on the classes and methods mentioned in this
section, see the following resources: 

- :java-sync-api:`bulkWrite() <com/mongodb/client/MongoCollection.html#bulkWrite(java.util.List,com.mongodb.client.model.BulkWriteOptions)>` API Documentation
- :java-core-api:`WriteModel <com/mongodb/client/model/WriteModel.html>` API Documentation
- :java-core-api:`BulkWriteOptions <com/mongodb/client/model/BulkWriteOptions.html>` API Documentation

Insert Operation
~~~~~~~~~~~~~~~~

To perform an insert operation, create an ``InsertOneModel`` specifying
the document you want to insert. To insert multiple documents, you must

> create an ``InsertOneModel`` for each document you want to insert.

Reviewer: This is great. It states how the user accomplishes a task, and answers the very natural follow-up question.

Example
```````

The following example creates an ``InsertOneModel`` for two documents

> where their ``_id`` are "3" and "4": 

Reviewer: suggestion: simplify this to the/is?

```suggestion
where the ``_id`` is "3" and "4": 
```

.. literalinclude:: /includes/fundamentals/code-snippets/BulkWrite.java
   :language: java
   :dedent:
   :start-after: begin insertDocumentsExample
   :end-before: end insertDocumentsExample

.. note::

   When performing a ``bulkWrite()``, the ``InsertOneModel`` cannot
   insert a document with an ``_id`` that already exists in the

>    collection. Instead, the method throws an ``MongoBulkWriteException``.

Reviewer: grammar: a/an

```suggestion
   collection. Instead, the method throws a ``MongoBulkWriteException``.
```

   The following example tries to insert two documents where the ``_id`` is
   "1" and "3":
   

>    .. literalinclude:: /includes/fundamentals/code-snippets/BulkWrite.java
>       :language: java
>       :dedent:
>       :start-after: begin insertExceptionExample
>       :end-before: end insertExceptionExample
> 
>    The following shows the output of the code above:
> 
>    .. code-block::
> 
>       Documents in the collection: 
>       {"_id": 1}
>       {"_id": 2}
> 
>       A MongoBulkWriteException occured: A MongoBulkWriteException occured with the following message:  
>       Bulk write operation error on server driver-ref-shard-00-02.pw0q4.mongodb.net:27017. 
>       Write errors: [BulkWriteError{index=0, code=11000, message='E11000 duplicate key error collection: 
>       crudOps.bulkWrite index: _id_ dup key: { _id: 1 }', details={}}].

Reviewer: suggestion: I like this note and example, but I'd remove the part where you log out the documents in the collection since that operation has nothing to do with the exception.

```suggestion
   .. literalinclude:: /includes/fundamentals/code-snippets/BulkWrite.java
      :language: java
      :dedent:
      :start-after: begin insertExceptionExample
      :end-before: end insertExceptionExample

   The following shows the output of the code above:

   .. code-block:: shell

      A MongoBulkWriteException occured: A MongoBulkWriteException occured with the following message:  
      Bulk write operation error on server driver-ref-shard-00-02.pw0q4.mongodb.net:27017. 
      Write errors: [BulkWriteError{index=0, code=11000, message='E11000 duplicate key error collection: 
      crudOps.bulkWrite index: _id_ dup key: { _id: 1 }', details={}}].
```

   To see why the document where the ``_id`` is "3" didn't insert, see
   the :ref:`Order of Execution <orderOfExecution>` section.

For additional information on the classes and methods mentioned in this
section, see the following resources: 

- :java-core-api:`InsertOneModel <com/mongodb/client/model/InsertOneModel.html>` API Documentation

Replace Operation
~~~~~~~~~~~~~~~~~

> To perform a replace operation, create an ``ReplaceOneModel`` specifying

Reviewer: grammar: a/an

```suggestion
To perform a replace operation, create a ``ReplaceOneModel`` specifying
```

the document you want to replace with the replacement document. 

.. note::

>    When performing a ``bulkWrite()``, the ``ReplaceOneModel`` cannot
>    make changes to a document that violate unique index constraints on
>    the collection. Also, the model does not replace a document if there
>    are no matches to your query filter. 

Reviewer: suggestion: The use of "also" makes it feel like the last sentence is an afterthought. Perhaps using "and" would be a better joiner?

```suggestion
   When performing a ``bulkWrite()``, the ``ReplaceOneModel`` cannot
   make changes to a document that violate unique index constraints on
   the collection, and the model does not replace a document if there
   are no matches to your query filter. 
```

Example
```````

> The following example creates a ``ReplaceOneModel`` to
> replace a document where the ``_id`` is "1" to a document with an
> additional field ``x`` with the value "4": 

Reviewer: suggestion: This reads strangely, as if you are describing an update operation "[update] to a document" rather than a replace operation "[replace] with a document". Also, unless the example is doing something with the field, I don't know if there's value in describing it. In the suggestion below I've omitted the information about the additional field to illustrate.

```suggestion
The following example creates a ``ReplaceOneModel`` to
replace a document where the ``_id`` is "1" with a document that
contains an additional field: 
```

.. literalinclude:: /includes/fundamentals/code-snippets/BulkWrite.java
   :language: java
   :dedent:
   :start-after: begin replaceDocumentsExample
   :end-before: end replaceDocumentsExample

For additional information on the classes and methods mentioned in this
section, see the following resources: 

- :java-core-api:`ReplaceOneModel <com/mongodb/client/model/ReplaceOneModel.html>` API Documentation
- :manual:`Unique indexes </core/index-unique/>` Server Manual Explanation

Update Operation
~~~~~~~~~~~~~~~~

> To perform an update operation, create an ``UpdateOneModel`` or an
> ``UpdateManyModel`` specifying the document you want to update with what
> the updates are.

Reviewer: suggestion: Use "query filter" here to make it more immediately clear.

```suggestion
To perform an update operation, create an ``UpdateOneModel`` or an
``UpdateManyModel`` specifying a query filter for documents you want to update with what
the updates are.
```

The ``UpdateOneModel`` updates one document that matches your query
filter and the ``UpdateManyModel`` updates all the documents that

> match your query filter. 

Reviewer: suggestion: clarify and expand. While correct it updates one document, does it update the `first` document that matches the query filter?

.. note::

   When performing a ``bulkWrite()``, the ``UpdateOneModel`` and
   ``UpdateManyModel`` cannot make changes to a document that violate
   unique index constraints on the collection. Also, the models do not

>    update any documents if there are no matches to your query filter. 

Reviewer: same as my previous suggestion about using "also" to open the last sentence

Example
```````

The following example creates an ``UpdateOneModel`` to update
a document where the ``_id`` is "2" to a document with an additional

> field ``x`` with the value "8": 

Reviewer: same as my previous suggestion about field details

.. literalinclude:: /includes/fundamentals/code-snippets/BulkWrite.java
   :language: java
   :dedent:
   :start-after: begin updateDocumentsExample
   :end-before: end updateDocumentsExample

For additional information on the classes and methods mentioned in this
section, see the following resources: 

- :java-core-api:`UpdateOneModel <com/mongodb/client/model/UpdateOneModel.html>` API Documentation
- :java-core-api:`UpdateManyModel <com/mongodb/client/model/UpdateManyModel.html>` API Documentation
- :manual:`unique indexes </core/index-unique/>` Server Manual Explanation

Delete Operation
~~~~~~~~~~~~~~~~

> To perform a delete operation, create a ``DeleteOneModel`` or a
> ``DeleteManyModel`` specifying the document you want to delete. 

Reviewer: suggestion: mention query filter here

```suggestion
To perform a delete operation, create a ``DeleteOneModel`` or a
``DeleteManyModel`` specifying a query filter that will match the documents 
you want to delete. 
```

The ``DeleteOneModel`` deletes one document that matches your query
filter and the ``DeleteManyModel`` deletes all the documents that
match your query filter. 

> 

Reviewer: same as before, is it more accurate to state it will delete the `first` document it matches?
